android/app/src/main/java/com/criticalblue/auth/demo/auth/AuthRepo.java

```diff
@@ -61,45 +63,92 @@
 import static com.criticalblue.auth.demo.auth.AuthEvent.AUTH_USER_INFO_FINISH;
 import static com.criticalblue.auth.demo.auth.AuthEvent.AUTH_USER_INFO_START;
 
+import javax.net.ssl.HttpsURLConnection;
+
 public class AuthRepo {
     private final String TAG = AuthRepo.class.getSimpleName();
 
     BooksApp app;
 
-    private Semaphore loginLock;
+    private final Semaphore loginLock;
 
     private AuthLoginListener loginListener;
-    private AuthLogoutListener logoutListener;
 
-    private AuthorizationService authService;
+    private final AuthorizationService authService;
     private AuthState authState;
     private String userInfoUrl;
+
     String clientId;
     String redirectUri;
     String authScope;
 
-    ConnectionBuilder connectionBuilder;
-
     public AuthRepo(BooksApp app) {
         this.app = app;
 
         loginLock = new Semaphore(1);
 
         loginListener = null;
-        logoutListener = null;
-
-        connectionBuilder = new SignedConnectionBuilder(app.getAssets());
-
-        AppAuthConfiguration.Builder builder = new AppAuthConfiguration.Builder();
-        builder.setConnectionBuilder(connectionBuilder);
-        authService = new AuthorizationService(app, builder.build());
         authState = null;
         userInfoUrl = null;
         clientId = null;
         redirectUri = null;
         authScope = null;
+
+        authService = new AuthorizationService(app, ApproovCustomConnection());
+    }
+
+    /**
+     * All the calls made by AppAuth will be secured by using the Approov Managed Trust Roots
+     * feature to protect the TLS channel against MitM attacks.
+     *
+     * @link https://approov.io/docs/latest/approov-usage-documentation/#managed-trust-roots
+     *
+     * @return The AppAuthConfiguration with a HttpsURLConnection secured by Approov
+     */
+    private AppAuthConfiguration ApproovCustomConnection() {
+        return new AppAuthConfiguration.Builder()
+                .setConnectionBuilder(new ConnectionBuilder() {
+                    @NonNull
+                    @Override
+                    public HttpsURLConnection openConnection(@NonNull Uri uri) throws IOException {
+                        URL url = new URL(uri.toString());
+                        HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
+
+                        io.approov.service.httpsurlconn.ApproovService.addApproov(connection);
+
+                        return connection;
+                    }
+                })
+                .build();
     }
 
+    /**
+     * Approov allows to fetch secrets just in time for being used to make API calls, provided that
```

Comment:
Maybe this should be clearer it initiates failLogin on a failure. Sounds very functional with no side effects as it is.

Reply:
I will add a note that will fail the login when an exception is thrown.

Reply:
Solved in the last commit.